### ISSUE TITLE
Error: Shared string not found for index

### ISSUE THREAD

**Issue by ankitpokhrel** [box/spout#6]
While reading excel sheet we are having a weird issue. We get the excel file from the external system. It runs fine if run unedited but throws following exception when we edit and save the same file. Even if we just add extra space or delete a sheet, the file throws the same exception.

> Shared string not found for index: 1432
#0 /opt/lampp/htdocs/mr/app/Vendor/Box/Spout/Reader/XLSX.php(247): Box\Spout\Reader\Helper\XLSX\SharedStringsHelper->getStringAtIndex(1432)
1 /opt/lampp/htdocs/mr/app/Vendor/Box/Spout/Reader/AbstractReader.php(131): Box\Spout\Reader\XLSX->read()

It displays random value which is not even in the current sheet. Note that we tested just by running the sample code provided under "How to read a XLSX file?" in documentation.

Any idea what may be going wrong?

**Comment by ankitpokhrel** [comment]
The problem is when there is multiline text in a cell. As you are exploding shared string file with `PHP_EOL` in [SharedStringHelper.php#243](https://github.com/box/spout/blob/master/src/Spout/Reader/Helper/XLSX/SharedStringsHelper.php#L243) it messes up the indexing when there is new line in the cell text itself.

We solved it temporarily by replacing `PHP_EOL` in text with unique character and then converting it back while retrieving the text.

Replace new line at [SharedStringHelper.php#122](https://github.com/box/spout/blob/master/src/Spout/Reader/Helper/XLSX/SharedStringsHelper.php#L122)
```
 $unescapedTextValue = str_replace(PHP_EOL, '&#10;', $unescapedTextValue);
```

and convert it back on retrieval. [SharedStringHelper.php#249](https://github.com/box/spout/blob/master/src/Spout/Reader/Helper/XLSX/SharedStringsHelper.php#L249)
```
$sharedString = str_replace('&#10;', PHP_EOL, $this->inMemoryTempFileContents[$indexInFile]);
```

But this doesn't seems to be the good solution. May be you should consider calculating and adding numeric index in the shared string file and make calculations based on those index.

**Comment by adrilo** [comment]
Thanks for reporting this issue @ankitpokhrel. And thanks for investigating!

Reading the temporary file containing the shared strings should remain a pretty fast task. Currently, as you pointed it out, it is assumed that the data is on one line. This makes the data retrieval really efficient because you know that the shared string with ID 85 will be on the 85th line.
Allowing strings to be on multiple lines will prevent us from making this assumption and lead to a more complex code. For instance, you could use `grep` to get the data that starts with "85--" (I tried using `grep` to parse the shared strings file - and avoid having to load the temp file in memory - but ended up not using because it's slow...). Another way to achieve this would be to load everything in memory as I do now, but instead of doing an explode, I could do something smarter like this:

Adding string index and special delimiters "||" around it at [SharedStringsHelper.php#L207](https://github.com/box/spout/blob/master/src/Spout/Reader/Helper/XLSX/SharedStringsHelper.php#L207)
```php
fwrite($this->tempFilePointer, "||$sharedStringIndex||$sharedString" . PHP_EOL);
```

Then, when loading the temp file in memory, instead of using explode -- [SharedStringsHelper.php#L243](https://github.com/box/spout/blob/master/src/Spout/Reader/Helper/XLSX/SharedStringsHelper.php#L243) -- I could have something like this:
```php
$fileHandle = fopen($tempFilePath, 'r');
if ($fileHandle) {
    $currentSharedStringIndex = 0;
    while (($line = fgets($fileHandle)) !== false) {
        if (preg_match('/\|\|(\d+)\|\|(.*)/', $line, $matches)) {
            $currentSharedStringIndex = $matches[1];
            $currentSharedString = $matches[2];
            $this->inMemoryTempFileContents[$currentSharedStringIndex] = $currentSharedString;
        } else {
            $this->inMemoryTempFileContents[$currentSharedStringIndex] .= PHP_EOL . $line;
        }
    }

    fclose($fileHandle);
}
```

And retrieve it that way:
```php
if (array_key_exists($sharedStringIndex, $this->inMemoryTempFileContents)) {
    $sharedString = $this->inMemoryTempFileContents[$sharedStringIndex];
}
```

This would still be pretty similar to your solution, as you need to determine a special delimiter that can be ignored on retrieval.

I can't think of another better solution that would not affect the complexity and performance of the code. So I'm going to implement your initial suggestion. If you have other suggestions, please let me know!

**Comment by ankitpokhrel** [comment]
Glad to know that you find the solution useful @adrilo . Another solution that I can think of is to use `special character + '\n' ` as the line separator for the plugin and explode file content based on this. Not sure how feasible it will be but that should work too.

**Comment by adrilo** [comment]
That would work but I like it less that the first solution you suggested. What I did is encode the \n character the way Excel does it. So the data remains valid and unchanged at any step (which is not true if an index or any extra character gets added).

I'll go ahead and merge my pull request.

**Comment by Alt0car** [comment]
The problem is always here ;
I'm trying toi parse an xlsx and i have this error :  Uncaught exception 'Box\Spout\Reader\Exception\SharedStringNotFoundException' with message 'Shared string not found for index: 2' ;

**Comment by adrilo** [comment]
@Alt0car which version of Spout are you using? The bug was fixed in 1.0.2.

**Comment by Alt0car** [comment]
I'm using the 1.0.2 bro ^^. So i think the bug are not really fixed...

**Comment by adrilo** [comment]
Interesting... @Alt0car If your XLSX file does not contain any sensitive data, would you mind sending it to me to help me debugging this issue? You can send it at adrien_at_box.com

**Comment by HKretschmer** [comment]
Same Problem here...fixing this would bring a huge advantage

**Comment by ankitpokhrel** [comment]
@Alt0car  and @HKretschmer  Are you guys working with multiple sheet or single sheet?

**Comment by HKretschmer** [comment]
@ankitpokhrel I'm using single sheet files

**Comment by ankitpokhrel** [comment]
@HKretschmer Strange...v1.0.3 is working fine for all of my sheets. Is it possible to send your excel file at info [at] ankitpokhrel [dot] com [dot] np ?

**Comment by HKretschmer** [comment]
@ankitpokhrel Sorry but those files contain sensible data...but I've found out, that there must be a context with empty cells...very strange. If I manually put some dummy data in those empty cells, save the sheet, delete the dummies and save the sheet again, it's working...has anybody an idea?

**Comment by adrilo** [comment]
@HKretschmer how did you create your XLSX? Through Excel or programmatically?
Also, can you confirm that I understand the problem correctly?

This does not work:

A | B | C
----- | ------ | ------
value A1 |  | value C1
value A2 | value B2 | value C2

But this does:

A | B | C
----- | ------ | ------
value A1 | _dummy value_ | value C1
value A2 | value B2 | value C2

Is that correct? Can you reproduce the issue with this setup?

**Comment by HKretschmer** [comment]
@adrilo The XLSX is exportet from another program (not manually created). The setup, that you have shown above describes the problem correctly. But usually I have no problems with empty cells.

**Comment by adrilo** [comment]
@HKretschmer so this error with empty cells only happens _sometimes_? Will there be a way for you to create a XLSX file without sensitive data where the issue occurs? I guess the XLSX file generated by your program slightly differs from what Spout expects, hence the error. Being able to reproduce the issue would be super useful!

**Comment by ankitpokhrel** [comment]
@adrilo I think the problem might be because of self closing tags which represent NO data instead of EMPTY data but I am not quite sure.
```
<si xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"><t/></si>
```

I received a sample file with issue from @HKretschmer sometime ago which I will forward you shortly.

**Comment by adrilo** [comment]
Thanks!

**Comment by adrilo** [comment]
@ankitpokhrel @HKretschmer Thank you for taking the time to reproduce the issue. I found the cause of it and have a fix for it: #21
